geiger/src/graph/NodeWithPort.js

```diff
@@ -0,0 +1,282 @@
+/**
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isObject} from 'util';
+import {
+  AUDIO_PARAM_RADIUS, INPUT_PORT_RADIUS, LEFT_SIDE_TOP_PADDING,
+  LEFT_TEXT_INDENT, TOTAL_INPUT_PORT_HEIGHT, TOTAL_PARAM_PORT_HEIGHT,
+} from '../ui/graphStyle';
+import {textSandbox} from '../ui/textSandbox';
+import Collection from '../util/Collection';
+import {computeAudioParamPortId, computeInPortId,
+  computeOutPortId} from './label';
+import {PortTypes} from './PortTypes';
+import Node from './Node';
+
+/**
+ * @typedef {Object} NodeCreationMessage
+ * @property {string} contextId
+ * @property {string} nodeId
+ * @property {string} nodeType
+ * @property {number} numberOfInputs
+ * @property {number} numberOfOutputs
+ * @property {Array<string>} audioParamNames
+ */
+
+/**
+ * @typedef {Object} NodeDimension
+ * @property {number} inputPortSectionHeight
+ * @property {number} paramPortSectionHeight
+ * @property {number} maxTextLength
+ * @property {number} totalHeight
+ */
+
+/**
+ * @typedef {Object} Port
+ * @property {string} id
+ * @property {PortTypes} type
+ * @property {string} text - The text label
+ * @property {number} y
+ */
+
+/**
+ * A node with audio params as ports.
+ * @class NodeWithPort
+ * @extends Node
+ */
+export default class NodeWithPort extends Node {
+  /**
+   * @constructor
+   * @param {!NodeCreationMessage} message
+   */
+  constructor(message) {
+    super(message);
+
+    this._ports = null;
+
+    this._initialize(message);
+  }
+
+  /**
+   * @param {!string} id
+   * @return {Port}
+   */
+  getPortById(id) {
+    return this._ports.get(id);
+  }
+
+  /**
+   * @param {!PortTypes} type
+   * @return {Array<Port>}
+   */
+  getPortsByType(type) {
+    return this._ports.filter((port) => port.type === type);
```

Comment:
nit: `(port)` can be just `port`.

Reply:
Linter complains if I remove the parentheses.